internal/pkg/controller/controller_test.go

```diff
@@ -1,254 +1,798 @@
 package controller
 
 import (
-	"math/rand"
+	"os"
 	"testing"
 	"time"
 
 	"github.com/sirupsen/logrus"
+	"github.com/stakater/Reloader/internal/pkg/common"
+	"github.com/stakater/Reloader/internal/pkg/testutil"
 	"github.com/stakater/Reloader/pkg/kube"
-	"k8s.io/api/core/v1"
-	"k8s.io/api/extensions/v1beta1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 )
 
 var (
+	client              = getClient()
+	namespace           = "test-reloader"
 	configmapNamePrefix = "testconfigmap-reloader"
 	secretNamePrefix    = "testsecret-reloader"
-	letters             = []rune("abcdefghijklmnopqrstuvwxyz")
+	data                = "dGVzdFNlY3JldEVuY29kaW5nRm9yUmVsb2FkZXI="
+	newData             = "dGVzdE5ld1NlY3JldEVuY29kaW5nRm9yUmVsb2FkZXI="
+	updatedData         = "dGVzdFVwZGF0ZWRTZWNyZXRFbmNvZGluZ0ZvclJlbG9hZGVy"
 )
 
-func randSeq(n int) string {
-	rand.Seed(time.Now().UnixNano())
-	b := make([]rune, n)
-	for i := range b {
-		b[i] = letters[rand.Intn(len(letters))]
+func TestMain(m *testing.M) {
+
+	logrus.Infof("Creating namespace %s", namespace)
+	testutil.CreateNamespace(namespace, client)
+
+	logrus.Infof("Creating controller")
+	for k := range kube.ResourceMap {
+		c, err := NewController(client, k, namespace)
+		if err != nil {
+			logrus.Fatalf("%s", err)
+		}
+
+		// Now let's start the controller
+		stop := make(chan struct{})
+		defer close(stop)
+		go c.Run(1, stop)
 	}
-	return string(b)
+	time.Sleep(5 * time.Second)
+
+	logrus.Infof("Running Testcases")
+	retCode := m.Run()
+
+	logrus.Infof("Deleting namespace %q.\n", namespace)
+	testutil.DeleteNamespace(namespace, client)
+
+	os.Exit(retCode)
 }
 
-// Creating a Controller to do a rolling upgrade upon updating the configmap or secret
+func getClient() *kubernetes.Clientset {
```

Comment:
Why is this duplicated in 2 test cases? Move it to testutil if its needed elsewhere